design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1067 @@
+packages in the cluster and refuse to create the `PackageRevision` if there is a
+known violation. To do so, the `Package` controller must be able to read and
+write the state of the packages it manages in a "concurrency-safe" manner. Let's
+look at an example.
+
+**Package A (PA) and Package B (PB) Both Want to Install CRD1 and CRD2**
+
+In the model established in Phase 1, this scenario could result in one of three
+outcomes:
+
+1. PA successfully installs and controls CRD1 and CRD2. It starts its packaged
+   controller and runs without issue.
+2. PB successfully installs and controls CRD1 and CRD2. It starts its packaged
+   controller and runs without issue.
+3. PB successfully installs and controls CRD1, but before it can install CRD2,
+   PA installs and controls it. Neither one of the package installations are
+   successful and none of the packaged controllers are started.
+
+While each of the controllers in this scenario are acting in a deterministic
+manner, the observed behavior is not consistent. We cannot alter this situation
+such that only PA or only PB will always win out and install successfully, but
+we can ensure that one or the other will install successfully (barring other
+violations).
+
+### PackageLock API
+
+The `PackageLock` type is a cluster-scoped singleton resource type that keeps a
+record of ownership and dependencies of Crossplane packages.
+
+```go
+// A PackageLock keeps track of Crossplane packages.
+type PackageLock struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   PackageLockSpec   `json:"spec,omitempty"`
+  Status PackageLockStatus `json:"status,omitempty"`
+}
+
+type PackageLockSpec struct {
+  Packages                  []PackageDependencies `json:"packages,omitempty"`
+  Compositions              map[string]string     `json:"compositions,omitempty"`
+  CustomResourceDefinitions map[string]string     `json:"crds,omitempty"`
+}
+
+type PackageDependencies struct {
+  Name         string   `json:"name,omitempty"`
+  Image        string   `json:"image,omitempty"`
+  Dependencies []string `json:"dependencies,omitempty"`
+}
+```
+
+### Package Controller Modifications
+
+With the introduction of the `PackageLock` type. The `Package` controller will
+be modified to both establish control of installed objects and check for missing
+dependencies prior to creating a new `PackageRevision`. It should report events
+and update the status of the `Package` to inform users of missing dependencies
+and conflicting ownership.
+
+### Action Items
+
+- [ ] Add `PackageLock` type
+- [ ] Update `Package` controller to handle dependency management
+- [ ] Deprecate `app-wordpress`
+- [ ] Update `stack-aws-sample` to use new `Package` format and composition
+- [ ] Update `stack-azure-sample` to use new `Package` format and composition
+- [ ] Update `stack-gcp-sample` to use new `Package` format and composition
```

Comment:
Why do these action items fall under this phase? Are they dependent on PackageLock?

Reply:
They are not strictly dependent on `PackageLock`, I think I just bumped them here to more evenly distribute work between phases.